File: server/nett.mjs
/* ============================================================
   Nett — de to utgående operasjonene importen trenger, i Node.

   Speiles av src-tauri/src/lib.rs for appmodus. Alt over dette
   laget er rene regler i src/importlogikk.mjs; her nede finnes
   bare henting, og vernet som må følge med den.
   ============================================================ */

import dns from "node:dns/promises";
import fs from "node:fs/promises";
import path from "node:path";
import { sjekkLenke } from "../src/felles.mjs";

const MAKS_SIDE   = 2 * 1024 * 1024;
const MAKS_HOPP   = 3;
const TIDSGRENSE  = 10_000;
const MODELL_URL  = "https://api.anthropic.com/v1/messages";
const NOKKELFIL   = "nokkel.txt";

const AGENT = "Jobbsoknader/1.0 (personlig soknadsoversikt)";

/* ---------- vern mot interne adresser ---------- */
/*
   Serveren henter en adresse brukeren limer inn, og står på samme
   maskin som datafilen. Uten denne sjekken er «importer denne
   utlysningen» en invitasjon til å lese http://127.0.0.1:4173/api/jobber,
   eller skymetadata på 169.254.169.254.

   Sjekken må gjøres på den oppslåtte IP-en, ikke på vertsnavnet:
   et navn i DNS kan peke hvor som helst. Og den må gjøres om igjen
   for hvert eneste hopp, ellers slipper en 302 forbi hele greia.
*/
export function erIntern(ip){
  const s = String(ip || "").trim().toLowerCase();
  if(!s) return true;

  /* IPv4-i-IPv6, «::ffff:10.0.0.1», gjelder som IPv4-en den bærer. */
  const kart = s.match(/^::ffff:(\d+\.\d+\.\d+\.\d+)$/);
  if(kart) return erIntern(kart[1]);

  const fire = s.match(/^(\d+)\.(\d+)\.(\d+)\.(\d+)$/);
  if(fire){
    const [a, b] = fire.slice(1).map(Number);
    if(a === 0 || a === 10 || a === 127) return true;          /* denne verten, privat, loopback */
    if(a === 169 && b === 254) return true;                     /* link-local og skymetadata */
    if(a === 172 && b >= 16 && b <= 31) return true;            /* privat */
    if(a === 192 && b === 168) return true;                     /* privat */
    if(a === 100 && b >= 64 && b <= 127) return true;           /* operatør-NAT */
    if(a >= 224) return true;                                   /* multicast og oppover */
    return false;
  }

  if(s === "::" || s === "::1") return true;                    /* uspesifisert, loopback */
  if(/^f[cd]/.test(s)) return true;                             /* unike lokale adresser */
  if(/^fe[89ab]/.test(s)) return true;                          /* link-local */
  return false;
}

async function sjekkVert(vert){
  /* Rå IP i adressen slipper ikke unna oppslaget — den er allerede svaret. */
  const bar = vert.replace(/^\[|\]$/g, "");
  if(/^[\d.]+$/.test(bar) || bar.includes(":")){
    if(erIntern(bar)) throw new Error("Adressen peker til et internt nett.");
    return;
  }
  let treff;
  try{ treff = await dns.lookup(bar, { all: true }); }
  catch{ throw new Error("Fant ikke serveren for den adressen."); }
  if(!treff.length || treff.some(t => erIntern(t.address)))
    throw new Error("Adressen peker til et internt nett.");
}

/* ---------- hent siden ---------- */
/*
   Redirects følges for hånd. `redirect: "follow"` ville gjort hoppene
   usynlige for sjekken over, og det er nettopp hoppene som er hullet.
*/
export async function hentSide(url){
  let na = url;

  for(let hopp = 0; hopp <= MAKS_HOPP; hopp++){
    const lenke = sjekkLenke(na);
    if(!lenke.ok) throw new Error(lenke.melding);

    const u = new URL(lenke.verdi);
    await sjekkVert(u.hostname);

    let r;
    try{
      r = await fetch(u, {
        redirect: "manual",
        signal: AbortSignal.timeout(TIDSGRENSE),
        headers: {
          "Accept": "text/html,application/xhtml+xml",
          "Accept-Language": "nb,no,en;q=0.8",
          "User-Agent": AGENT
        }
      });
    }catch(e){
      throw new Error(e?.name === "TimeoutError"
        ? "Siden svarte ikke innen ti sekunder."
        : "Fikk ikke kontakt med siden.");
    }

    if(r.status >= 300 && r.status < 400 && r.headers.get("location")){
      na = new URL(r.headers.get("location"), u).href;
      continue;
    }

    if(!r.ok) throw new Error(`Siden svarte ${r.status}.`);

    const type = (r.headers.get("content-type") || "").toLowerCase();
    if(type && !type.includes("html") && !type.includes("xml") && !type.includes("text/plain"))
      throw new Error("Adressen peker ikke til en nettside.");

    return { status: r.status, sluttUrl: u.href, html: await lesMedTak(r) };
  }

  throw new Error("Siden sendte oss videre for mange ganger.");
}

/* Content-Length kan lyve; taket må gjelde det som faktisk kommer inn. */
async function lesMedTak(r){
  if(!r.body) return "";
  const leser = r.body.getReader();
  const biter = [];
  let n = 0;
  for(;;){
    const { done, value } = await leser.read();
    if(done) break;
    n += value.length;
    if(n > MAKS_SIDE){ await leser.cancel(); break; }
    biter.push(value);
  }
  return new TextDecoder("utf-8").decode(Buffer.concat(biter.map(Buffer.from)));
}

/* ---------- spør modellen ---------- */

export class ManglerNokkel extends Error {
  constructor(melding){ super(melding); this.navn = "mangler-nokkel"; }
}

/* Nøkkelen ligger helst i en fil ved siden av datafilen: den kan settes
   chmod 600, den er allerede i .gitignore, og den slipper å stå i klartekst
   i ~/.zshrc der hvert eneste program på maskinen ser den. Miljøvariabelen
   virker fortsatt og går foran, for den som vil ha den.

   Filen leses ved hvert kall. Det er én liten lesing per import, og til
   gjengjeld virker en ny nøkkel uten at serveren må startes på nytt. */
export async function lesNokkel(katalog){
  const fra = process.env.ANTHROPIC_API_KEY?.trim();
  if(fra) return fra;

  if(katalog){
    try{
      const t = (await fs.readFile(path.join(katalog, NOKKELFIL), "utf8")).trim();
      if(t) return t;
    }catch{ /* finnes ikke, eller kan ikke leses — samme svar som ingen nøkkel */ }
  }

  throw new ManglerNokkel(
    `Ingen API-nøkkel. Legg den i ${NOKKELFIL} i datakatalogen, eller sett ANTHROPIC_API_KEY.`);
}

/* Samler de to operasjonene slik lagLager samler filoperasjonene, og
   av samme grunn: katalogen er noe serveren vet, ikke noe modulen gjør. */
export function lagNett({ katalog } = {}){
  return { hentSide, spørModell: kropp => spørModell(kropp, katalog) };
}

/* Verdt ett nytt forsøk: køen er full, eller noe er nede et øyeblikk.
   Uten dette må brukeren trykke selv — og da betales hentingen på nytt. */
const PRØV_IGJEN = new Set([408, 409, 429, 500, 502, 503, 504, 529]);
const PAUSE = 1_500;

const sov = ms => new Promise(r => setTimeout(r, ms));

export async function spørModell(kropp, katalog){
  const nokkel = await lesNokkel(katalog);

  const send = () => fetch(MODELL_URL, {
    method: "POST",
    signal: AbortSignal.timeout(30_000),
    headers: {
      "x-api-key": nokkel,
      "anthropic-version": "2023-06-01",
      "content-type": "application/json"
    },
    body: JSON.stringify(kropp)
  });

  let r;
  try{
    r = await send();
    if(PRØV_IGJEN.has(r.status)){
      /* Retry-After er serverens eget råd; den vet bedre enn vi gjør. */
      const raad = Number(r.headers.get("retry-after"));
      await sov(Number.isFinite(raad) && raad > 0 ? Math.min(raad * 1000, 10_000) : PAUSE);
      r = await send();
    }
  }catch(e){
    throw new Error(e?.name === "TimeoutError"
      ? "Modellen svarte ikke i tide."
      : "Fikk ikke kontakt med modellen.");
  }

  const tekst = await r.text();
  if(!r.ok){
    /* Nøkkelen skal aldri kunne havne i en feilmelding brukeren ser. */
    let melding = `Modellen svarte ${r.status}.`;
    try{
      const j = JSON.parse(tekst);
      if(j?.error?.message) melding = `Modellen svarte ${r.status}: ${j.error.message}`;
    }catch{ /* behold den enkle meldingen */ }
    throw new Error(melding);
  }

  try{ return JSON.parse(tekst); }
  catch{ throw new Error("Modellen svarte noe annet enn JSON."); }
}

File: src-tauri/src/nett.rs
/* ============================================================
   De to utgående operasjonene, i appmodus.

   Motstykket til server/nett.mjs: samme to operasjoner, samme
   vern, men utført av Rust. Reglene over dette laget ligger i
   src/importlogikk.mjs og deles med nettleserversjonen.

   Nøkkelen leses her og forlater aldri Rust. Adressen til modellen
   står i koden, ikke i et argument — ellers ville en fiendtlig
   utlysning kunnet be oss sende nøkkelen et annet sted.
   ============================================================ */

use std::net::{IpAddr, SocketAddr, ToSocketAddrs};
use std::time::Duration;

use serde::Serialize;

const MAKS_SIDE: usize = 2 * 1024 * 1024;
const MAKS_HOPP: usize = 3;
const TIDSGRENSE: Duration = Duration::from_secs(10);
const TIDSGRENSE_MODELL: Duration = Duration::from_secs(30);
const MODELL_URL: &str = "https://api.anthropic.com/v1/messages";
const AGENT: &str = "Jobbsoknader/1.0 (personlig soknadsoversikt)";

#[derive(Serialize, Debug)]
pub struct Side {
    pub status: u16,
    #[serde(rename = "sluttUrl")]
    pub slutt_url: String,
    pub html: String,
}

/// Adresser vi aldri skal hente fra. Brukeren limer inn lenken, og
/// appen kjører på samme maskin som datafilen: uten denne sjekken er
/// «importer denne utlysningen» en måte å be appen lese det som ligger
/// på loopback eller på skymetadata-adressen.
pub fn er_intern(ip: &IpAddr) -> bool {
    match ip {
        IpAddr::V4(v4) => {
            v4.is_loopback()
                || v4.is_private()
                || v4.is_link_local()
                || v4.is_broadcast()
                || v4.is_documentation()
                || v4.is_multicast()
                || v4.is_unspecified()
                || v4.octets()[0] == 0
                // operatør-NAT, 100.64.0.0/10
                || (v4.octets()[0] == 100 && (64..=127).contains(&v4.octets()[1]))
        }
        IpAddr::V6(v6) => {
            if let Some(kart) = v6.to_ipv4_mapped() {
                return er_intern(&IpAddr::V4(kart));
            }
            v6.is_loopback()
                || v6.is_multicast()
                || v6.is_unspecified()
                // unike lokale adresser, fc00::/7
                || (v6.segments()[0] & 0xfe00) == 0xfc00
                // link-local, fe80::/10
                || (v6.segments()[0] & 0xffc0) == 0xfe80
        }
    }
}

/// Slår opp verten og godtar den bare hvis hver eneste adresse er
/// utvendig. Den kontrollerte adressen sendes tilbake, slik at
/// klienten kan pinnes til den — ellers kunne oppslaget gitt ett svar
/// til sjekken og et annet til koblingen.
fn sjekk_vert(vert: &str, port: u16) -> Result<SocketAddr, String> {
    let treff: Vec<SocketAddr> = (vert, port)
        .to_socket_addrs()
        .map_err(|_| "Fant ikke serveren for den adressen.".to_string())?
        .collect();

    if treff.is_empty() {
        return Err("Fant ikke serveren for den adressen.".into());
    }
    if treff.iter().any(|a| er_intern(&a.ip())) {
        return Err("Adressen peker til et internt nett.".into());
    }
    Ok(treff[0])
}

/// Redirects følges for hånd. Lot vi reqwest gjøre det, ville hoppene
/// vært usynlige for sjekken over — og det er hoppene som er hullet.
pub async fn hent_side(url: String) -> Result<Side, String> {
    let mut na = url;

    for _ in 0..=MAKS_HOPP {
        let u = reqwest::Url::parse(&na).map_err(|_| {
            "Lenken må være en full nettadresse, f.eks. https://…".to_string()
        })?;
        if u.scheme() != "http" && u.scheme() != "https" {
            return Err("Bare http- og https-lenker er tillatt.".into());
        }
        let vert = u.host_str().ok_or("Lenken mangler et vertsnavn.")?.to_string();
        let port = u.port_or_known_default().unwrap_or(443);
        let adresse = sjekk_vert(&vert, port)?;

        let klient = reqwest::Client::builder()
            .redirect(reqwest::redirect::Policy::none())
            .timeout(TIDSGRENSE)
            .user_agent(AGENT)
            .resolve(&vert, adresse)
            .build()
            .map_err(|e| format!("Fikk ikke satt opp hentingen: {e}"))?;

        let svar = klient
            .get(u.clone())
            .header("Accept", "text/html,application/xhtml+xml")
            .header("Accept-Language", "nb,no,en;q=0.8")
            .send()
            .await
            .map_err(|e| {
                if e.is_timeout() {
                    "Siden svarte ikke innen ti sekunder.".to_string()
                } else {
                    "Fikk ikke kontakt med siden.".to_string()
                }
            })?;

        let kode = svar.status();
        if kode.is_redirection() {
            if let Some(sted) = svar.headers().get(reqwest::header::LOCATION) {
                let sted = sted.to_str().map_err(|_| "Ugyldig videresending.".to_string())?;
                na = u.join(sted).map_err(|_| "Ugyldig videresending.".to_string())?.into();
                continue;
            }
        }
        if !kode.is_success() {
            return Err(format!("Siden svarte {}.", kode.as_u16()));
        }

        let type_ = svar
            .headers()
            .get(reqwest::header::CONTENT_TYPE)
            .and_then(|v| v.to_str().ok())
            .unwrap_or("")
            .to_ascii_lowercase();
        if !type_.is_empty()
            && !type_.contains("html")
            && !type_.contains("xml")
            && !type_.contains("text/plain")
        {
            return Err("Adressen peker ikke til en nettside.".into());
        }

        return Ok(Side {
            status: kode.as_u16(),
            slutt_url: u.to_string(),
            html: les_med_tak(svar).await?,
        });
    }

    Err("Siden sendte oss videre for mange ganger.".into())
}

/// Content-Length kan lyve; taket må gjelde det som faktisk kommer inn.
async fn les_med_tak(mut svar: reqwest::Response) -> Result<String, String> {
    let mut biter: Vec<u8> = Vec::new();
    while let Some(bit) = svar
        .chunk()
        .await
        .map_err(|_| "Mistet forbindelsen mens siden ble lest.".to_string())?
    {
        if biter.len() + bit.len() > MAKS_SIDE {
            biter.extend_from_slice(&bit[..MAKS_SIDE - biter.len()]);
            break;
        }
        biter.extend_from_slice(&bit);
    }
    Ok(String::from_utf8_lossy(&biter).into_owned())
}

fn prøv_igjen(kode: u16) -> bool {
    matches!(kode, 408 | 409 | 429 | 500 | 502 | 503 | 504 | 529)
}

/// Kroppen kommer som tekst fra JavaScript; adressen og nøkkelen gjør
/// ikke det. Nøkkelen returneres aldri, heller ikke i en feilmelding.
pub async fn spor_modell(nokkel: String, kropp: String) -> Result<String, String> {
    let klient = reqwest::Client::builder()
        .timeout(TIDSGRENSE_MODELL)
        .user_agent(AGENT)
        .build()
        .map_err(|e| format!("Fikk ikke satt opp forespørselen: {e}"))?;

    let send = || {
        klient
            .post(MODELL_URL)
            .header("x-api-key", &nokkel)
            .header("anthropic-version", "2023-06-01")
            .header("content-type", "application/json")
            .body(kropp.clone())
            .send()
    };

    let feilmelding = |e: reqwest::Error| {
        if e.is_timeout() {
            "Modellen svarte ikke i tide.".to_string()
        } else {
            "Fikk ikke kontakt med modellen.".to_string()
        }
    };

    let mut svar = send().await.map_err(feilmelding)?;

    /* Ett nytt forsøk når køen er full eller noe er nede et øyeblikk.
       Uten det må brukeren trykke selv, og betaler hentingen på nytt. */
    if prøv_igjen(svar.status().as_u16()) {
        let pause = svar
            .headers()
            .get(reqwest::header::RETRY_AFTER)
            .and_then(|v| v.to_str().ok())
            .and_then(|v| v.parse::<u64>().ok())
            .map(|s| Duration::from_secs(s.min(10)))
            .unwrap_or(Duration::from_millis(1500));
        tokio::time::sleep(pause).await;
        svar = send().await.map_err(feilmelding)?;
    }

    let kode = svar.status();
    let tekst = svar
        .text()
        .await
        .map_err(|_| "Klarte ikke lese svaret fra modellen.".to_string())?;

    if !kode.is_success() {
        let melding = serde_json::from_str::<serde_json::Value>(&tekst)
            .ok()
            .and_then(|j| j["error"]["message"].as_str().map(str::to_owned))
            .map(|m| format!("Modellen svarte {}: {m}", kode.as_u16()))
            .unwrap_or_else(|| format!("Modellen svarte {}.", kode.as_u16()));
        return Err(melding);
    }
    Ok(tekst)
}

#[cfg(test)]
mod tester {
    use super::*;

    fn ip(s: &str) -> IpAddr {
        s.parse().expect("gyldig ip i testen")
    }

    #[test]
    fn interne_adresser_avvises() {
        for a in [
            "127.0.0.1",
            "10.1.2.3",
            "172.16.0.1",
            "172.31.255.254",
            "192.168.0.5",
            "169.254.169.254", // skymetadata
            "0.0.0.0",
            "100.64.0.1",      // operator-NAT
            "224.0.0.1",
            "::1",
            "::",
            "fe80::1",
            "fd00::1",
            "::ffff:127.0.0.1", // IPv4 i IPv6-drakt
            "::ffff:10.0.0.1",
        ] {
            assert!(er_intern(&ip(a)), "{a} skulle vaert avvist");
        }
    }

    #[test]
    fn utvendige_adresser_slipper_gjennom() {
        for a in [
            "93.184.216.34",
            "8.8.8.8",
            "172.32.0.1", // rett utenfor 172.16/12
            "99.64.0.1",  // rett utenfor 100.64/10
            "2606:2800:220:1:248:1893:25c8:1946",
        ] {
            assert!(!er_intern(&ip(a)), "{a} skulle sluppet gjennom");
        }
    }

    #[test]
    fn bare_forbigående_feil_gir_nytt_forsøk() {
        for k in [408, 429, 500, 502, 503, 529] {
            assert!(prøv_igjen(k), "{k} skulle gitt nytt forsøk");
        }
        for k in [200, 400, 401, 403, 404, 422] {
            assert!(!prøv_igjen(k), "{k} skal ikke prøves på nytt");
        }
    }

    #[tokio::test]
    async fn loopback_hentes_ikke() {
        let feil = hent_side("http://127.0.0.1:4173/api/jobber".into())
            .await
            .expect_err("loopback skal avvises");
        assert!(feil.contains("internt nett"), "uventet melding: {feil}");
    }

    #[tokio::test]
    async fn bare_http_og_https() {
        for url in ["file:///etc/passwd", "javascript:alert(1)", "ftp://example.com/"] {
            assert!(hent_side(url.into()).await.is_err(), "{url} skulle vaert avvist");
        }
    }
}
